Warn of standard behavior of 'default' key store

tdietrich513 opened the pull request:
+label: docsite_pr

##### SUMMARY
Add some verbiage to the documentation to help users avoid a cryptic error.

##### ISSUE TYPE
- Docs Pull Request

##### COMPONENT NAME
win_certificate_store

##### ADDITIONAL INFORMATION
The default key store typically results in the key being stored as user. In instances where we're trying to install a cert to enable SSL binding in IIS, this will result in an unbind-able cert, as the IIS APPPOOL identity will not be able to read the private key, and presents a very cryptic error of 'A specified logon session does not exist'. Setting the key_store to machine avoids this error and this concept should be represented in the documentation for this very common use case.

tdietrich513 committed:
Warn of standard behavior of 'default' key store 

The default key store typically results in the key being stored as user. In instances where we're trying to install a cert to enable SSL binding in IIS, this will result in an unbind-able cert, as the IIS APPPOOL identity will not be able to read the private key, and presents a very cryptic error of 'A specified logon session does not exist'.


+label: docsite_pr

tdietrich513 committed:
fix trailing whitespace

JohnLieske commented:
Looks like a legitimate clarification.

jborean93 commented on lib/ansible/modules/windows/win_certificate_store.py:
```diff
@@ -177,6 +181,15 @@
+    path: C:\Temp\cert.pfx
+    file_type: pkcs12
+    password: StrongPassword!
+    key_exportable: yes
```
I thought that IIS didn't require the key to be exportable, just that a key is imported and usable.

tdietrich513 replied:
You are correct that it is not necessary but it is generally recommended in most of the documentation I've read.

jborean93 replied:
I just feel it's more of a security issue, I prefer to document the absolute minimum that is needed and not have people open themselves up more.

tdietrich513 replied:
Fair enough. I've removed the line.

jborean93 commented on lib/ansible/modules/windows/win_certificate_store.py:
```diff
@@ -91,7 +91,7 @@
     description:
     - Specifies where Windows will store the private key when it is imported.
-    - When set to C(default), the default option as set by Windows is used.
+    - When set to C(default), the default option as set by Windows is used- typically user.
```
probably best to turn the `-` into `,` and enclose user with `C(user)`.

tdietrich513 replied:
Done.

jborean93 commented on lib/ansible/modules/windows/win_certificate_store.py:
```diff
@@ -118,6 +118,10 @@
 - The certificates must be located on the Windows host to be set with I(path).
+- When importing a certificate for usage in IIS, it is generally required
+  to use the C(machine) key_storage option, as both C(default) and C(user)
+  will make the private key unreadable to IIS APPPOOL identities and prevent
```
It's less the private key but the whole certificate and key not being accessible to that account.

tdietrich513 replied:
Not sure this is correct unless you use the `CurrentUser` certificate store. `LocalMachine` is the default. In the situation that lead me to suggest this documentation change, I was able to view and select the certificate from a different user account without issue, but IIS would fail in attempting to bind the certificate to the HTTPS webbinding because it was unable to access the private key.

jborean93 replied:
> LocalMachine is the default

I disagree, we have seen how the default option on Windows is not actually LocalMachine, this is the whole point of this PR to point out how the default may not be what you need :)

> I was able to view and select the certificate from a different user account without issue

I'm not able to do so, when I go to select a certificate I am only able to see the certs in the Personal store of LocalMachine, any certs in the Personal store are not viewable and therefore not selectable.

tdietrich513 replied:
I think we're talking about two different things, namely Certificate Store Locations versus Private Key Storage. This doc PR is about the PK storage, not the certificate store.

`LocalMachine` is the default Certificate Store Location in the script this module uses, while `Default` is the default private key storage, which may resolve to the `User` private key storage OR the `Machine` private key storage, depending on what the default key storage is configured to be in windows (about which microsoft's documentation only says "Generally this will be `User`"). 

It's entirely expected that if you install a certificate into the `CurrentUser` certificate store it would only be available to whatever account Ansible is using while running the install.

tdietrich513 replied:
I went ahead and added `store_location: LocalMachine` to the example, I agree that it is probably a good idea not to have the documentation rely on the default setting to be correct for the use case as the default could change.

jborean93 replied:
Ahh sorry, complete mind blank there and totally forgot we are talking about `key_storage` and not `store_location`. Please ignore this comment.

tdietrich513 committed:
Use C(user) reference

tdietrich513 committed:
specify what's required for successful iis import

- store_location's default value is currently 'LocalMachine' but if this changes it could result in the documentation being inaccurate
- key_exportable is not required for a successful IIS import

acozine commented on lib/ansible/modules/windows/win_certificate_store.py:
```diff
@@ -38,7 +38,7 @@
     - The path to a certificate file.
     - This is required when I(state) is C(present) or C(exported).
-    - When I(state) is C(absent) and I(thumbprint) is not specified, the
+    - When I(state) is C(absent) and I(thumbprint) is not specified, the=
```
```suggestion
    - When I(state) is C(absent) and I(thumbprint) is not specified, the
```

acozine replied:
I think the stray `=` was probably a typo . . . or a cat on your keyboard.

tdietrich513 replied:
Dangit 🐱

acozine committed:
remove stray =

dang cat

Co-Authored-By: tdietrich513 <tom.dietrich@gmail.com>

jborean93 commented:
Thanks @tdietrich513 for your work on this.

jborean93 commented:
It would be great if you could backport this PR to stable-2.7 so it is updated in the latest docs, a guide on how to do this can be found here https://docs.ansible.com/ansible/latest/community/development_process.html#backport-pull-request-process.
